### DISCUSSION TITLE
VICI 'child-updown' event message not clear

### DISCUSSION THREAD

**Discussion by vishal1kumar** [strongswan/strongswan#540] [Q&A] [answered]
Hi Community,

I am writing an vici event handler where I have registered with vici for 'child-updown' events. 
In general cases where the tunnel is going up or down (due to termination request from peer), vici message for 'child-updown' is clear, they clearly mention ('up', 'yes') ,   ('state', 'ESTABLISHED') OR ('state', 'DELETING') .

I am confused with vici message for chile_sa delete message in cases of 'INITIAL_CONTACT' and DPD expire. In these cases vici is sending a message which doesnt specify the state as DELETED/DELETING. The message looks like below:
`OrderedDict([('3412721420_ipsec_loc_1_50002', OrderedDict([('uniqueid', b'10'), ('version', b'2'), ('state', b'ESTABLISHED'), ('local-host', b'172.17.0.7'), ('local-port', b'4500'), ('local-id', b'10.140.208.194'), ('remote-host', b'10.140.216.53'), ('remote-port', b'4500'), ('remote-id', b'10.140.216.53'), ('initiator-spi', b'5c5735ca5dcb5ef6'), ('responder-spi', b'b49aaf41ca7d1f31'), ('nat-local', b'yes'), ('nat-any', b'yes'), ('encr-alg', b'AES_CBC'), ('encr-keysize', b'128'), ('integ-alg', b'HMAC_SHA1_96'), ('prf-alg', b'PRF_HMAC_SHA1'), ('dh-group', b'MODP_2048'), ('established', b'255'), ('rekey-time', b'82808'), ('tasks-active', [b'IKE_DPD']), ('child-sas', OrderedDict([('child-sa-3', OrderedDict([('name', b'child-sa'), ('uniqueid', b'3'), ('reqid', b'2'), ('state', b'INSTALLED'), ('mode', b'TUNNEL'), ('protocol', b'ESP'), ('encap', b'yes'), ('spi-in', b'c516e6e9'), ('spi-out', b'c810a894'), ('mark-in', b'0000c352'), ('mark-out', b'0000c352'), ('encr-alg', b'AES_CBC'), ('encr-keysize', b'128'), ('integ-alg', b'HMAC_SHA1_96'), ('bytes-in', b'0'), ('packets-in', b'0'), ('bytes-out', b'0'), ('packets-out', b'0'), ('rekey-time', b'3242'), ('life-time', b'3705'), ('install-time', b'255'), ('local-ts', [b'0.0.0.0/0[tcp/ftp]', b'0.0.0.0/0[tcp/http]', b'0.0.0.0/0[tcp/https]']), ('remote-ts', [b'10.140.216.0/24'])]))]))]))])`

My question is:
**With above vici message, how can I identify that the child_sa is deleted in charon? 
Are there any other message which I need to parse and handle properly where I may miss to handle in my vici_event_handler?**

Thanks

**Comment by tobiasbrunner** [comment, marked as answer]
If there is no `up` key, it means it's a "down" event (it's only added for "up" events and set to "yes" there). That's the same with pretty much all boolean flags sent via vici.

Also note that a CHILD_SA is _never_ in state `ESTABLISHED` that's a state only available for IKE_SAs (proper CHILD_SAs are in state `INSTALLED` as can be seen here).

**Reply by vishal1kumar** [reply]
Thanks Tobias this clarifies.
